### PR TITLE
write gift notification date to zuora, when DS is purchased

### PR DESCRIPTION
**Pull request by johnduffell**
## What are you doing in this PR?

This PR makes the step functions write the gift notification date to zuora when a DS gift is purchased.

This field will end up in the DL, so we can do a one off report on this in a couple of weeks.

https://trello.com/c/gRopJsmH/3498-write-gift-delivery-date-to-zuora

## Why are you doing this?

CEX team are concerned that if a lot of gifts are scheduled for e.g. christmas day, it might cause a risk to support, as the call centre is not open over christmas/boxing day.

## Screenshots

Tested in DEV:
https://apisandbox.zuora.com/apps/Subscription.do?method=view&id=2c92c0f975e5ba1e01761e9e1109498d
image

Tested in UAT:
https://apisandbox.zuora.com/apps/Subscription.do?method=view&id=2c92c0f975e5ba2601761e9f907d5bcb
image

I also tested a normal Direct DS in UAT.

PROD field name:
image

### CODE REVIEW

File: support-models/src/main/scala/com/gu/support/zuora/api/SubscriptionData.scala [hunk 1]
```diff
@@ -179,7 +179,8 @@ case class Subscription(
   readerType: ReaderType = ReaderType.Direct,
   promoCode: Option[PromoCode] = None,
   redemptionCode: Option[RawRedemptionCode] = None,
-  corporateAccountId: Option[String] = None
+  corporateAccountId: Option[String] = None,
+  giftNotificationEmailDate__c: Option[LocalDate] = None,
```
Comment (rupertbates) [inline support-models/src/main/scala/com/gu/support/zuora/api/SubscriptionData.scala:183]: Can we remove the `__c` from the end of this field and map it as we do with the other custom fields on this object here: https://github.com/guardian/support-frontend/blob/8cd4badd66b080321f84f79147459e1351ed48c0/support-models/src/main/scala/com/gu/support/zuora/api/SubscriptionData.scala#L165-L164

File: support-workers/src/main/scala/com/gu/emailservices/EmailFields.scala [hunk 2]
```diff
@@ -50,7 +50,7 @@ case class EmailFields(
       DataExtensionName = dataExtensionName,
       SfContactId = userId.left.toOption.map(_.id),
       IdentityUserId = userId.right.toOption.map(_.id),
-      deliveryDate.map(_.toDateTime(new LocalTime(8, 0), DateTimeZone.UTC)),
+      deliveryDate.map(_.toDateTime(LocalTime.now().plusSeconds(90))),//new LocalTime(8, 0), DateTimeZone.UTC)),
```
Comment (rupertbates) [inline support-workers/src/main/scala/com/gu/emailservices/EmailFields.scala:53]: Is this change intentional?

Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/emailservices/EmailFields.scala:53]: no good spot, i thought I unticked that one

File: support-models/src/main/scala/com/gu/support/zuora/api/SubscriptionData.scala (hunk 1 above)
Comment (johnduffell) [reply support-models/src/main/scala/com/gu/support/zuora/api/SubscriptionData.scala:183]: I was wondering that, we should probably do the same for createdRequestId__c: String, on line 173 too.

Comment (johnduffell) [reply support-models/src/main/scala/com/gu/support/zuora/api/SubscriptionData.scala]: I've removed it from both now so it should be consistent.

File: support-models/src/main/scala/com/gu/support/zuora/api/SubscriptionData.scala [hunk 3]
```diff
@@ -152,25 +152,23 @@ object SubscriptionProductFeature {
 case class SubscriptionProductFeature(featureId: String)
 
 object Subscription {
-  implicit val decoder: Decoder[Subscription] = decapitalizingDecoder[Subscription].prepare(
-    _.withFocus(_.mapObject(_.renameField("PromotionCode__c", "promoCode")))
-      .withFocus(_.mapObject(_.renameField("ReaderType__c", "readerType")))
-  )
```
Comment (johnduffell) [inline support-models/src/main/scala/com/gu/support/zuora/api/SubscriptionData.scala:158]: this was out of date and we should never need to deserialise this as it's part of the request not the response.
I removed it completely and turned all the codecs that depend on it into Encoders and the maain code still seemed to compile okay.

File: support-models/src/test/scala/com/gu/support/zuora/api/SerialisationSpec.scala [hunk 4]
```diff
@@ -36,17 +36,71 @@ class SerialisationSpec extends AsyncFlatSpec with SerialisationTestHelpers with
...
+              |    "InitialTerm" : 3,
+              |    "RenewalTerm" : 12,
+              |    "TermType" : "TERMED",
+              |    "ReaderType__c" : "Gift",
+              |    "RedemptionCode__c" : "gd03-asdfghjq",
+              |    "CreatedRequestId__c" : "requestId",
+              |    "GiftNotificationEmailDate__c" : "2020-12-25"
+              |  }
+              |}""".stripMargin).toTry.get
+    json.mapObject(_.mapValues(_.dropNullValues)) should be(expected)
```
Comment (johnduffell) [inline support-models/src/test/scala/com/gu/support/zuora/api/SerialisationSpec.scala:94]: getting rid of the nulls was tricky - we drop them in the json printer
https://github.com/guardian/support-frontend/blob/547ca4ea72ba19598165c89abb443ec289b5d9de/support-modules/rest/src/main/scala/com/gu/rest/WebServiceHelper.scala#L142
 but if we only go as far as the json, that doesn't apply.

File: support-workers/src/main/scala/com/gu/zuora/subscriptionBuilders/DigitalSubscriptionPurchaseBuilder.scala [hunk 5]
```diff
@@ -28,15 +28,16 @@ class DigitalSubscriptionPurchaseBuilder(
...
-    val (contractAcceptanceDelay, autoRenew, initialTerm, maybeRedemptionCode) =
-      (digitalPack.readerType, maybeGiftCode) match {
-        case (Gift, Some(giftCode)) => (0, false, GiftCodeValidator.expirationTimeInMonths + 1, Some(giftCode))
-        case (Gift, _) | (_, Some(_)) => throw new RuntimeException("coding error - possible unredeemable sub")
-        case _ => (config.defaultFreeTrialPeriod + config.paymentGracePeriod, true, 12, None)
+    val (contractAcceptanceDelay, autoRenew, initialTerm, maybeRedemptionCode, deliveryDate) =
+      (digitalPack.readerType, maybeGiftCode, maybeDeliveryDate) match {
+        case (Gift, Some(giftCode), Some(deliveryDate)) => (0, false, GiftCodeValidator.expirationTimeInMonths + 1, Some(giftCode), Some(deliveryDate))
+        case (Gift, _, _) | (_, Some(_), _) | (Gift, _, None) => throw new RuntimeException("coding error - possible unredeemable sub")
+        case _ => (config.defaultFreeTrialPeriod + config.paymentGracePeriod, true, 12, None, None)
```
Comment (johnduffell) [inline support-workers/src/main/scala/com/gu/zuora/subscriptionBuilders/DigitalSubscriptionPurchaseBuilder.scala:40]: now it will blow up if we get a Gift but we don't have a delivery date to put into zuora, otherwise we might not notice.  But it does mean that if we break that, all gift checkouts will fail, rather than going through with missing data.
If we think that's too harsh but still want some checking, we could make it a metric.
Alternatively, once we finish refactoring the state types, this is the sort of thing that would be checked by the type system anyway.

Comment (rupertbates) [reply support-workers/src/main/scala/com/gu/zuora/subscriptionBuilders/DigitalSubscriptionPurchaseBuilder.scala:40]: I think that's fine

**Review by rupertbates — APPROVED** [review]
Good work 👍
